skimage/restoration/_cycle_spin.py

```diff
@@ -0,0 +1,181 @@
+        ``multiprocessing.cpu_count`` will be used.
+    multichannel : bool, optional
+        Whether to treat the final axis as channels (no cycle shifts are
+        performed over the channels axis).
+    func_kw : dict, optional
+        Additional keyword arguments to supply to ``func``.
+
+    Returns
+    -------
+    avg_y : np.ndarray
+        The output of ``func(x, **func_kw)`` averaged over all combinations of
+        the specified axis shifts.
+
+    Notes
+    -----
+    Cycle spinning was proposed as a way to approach shift-invariance via
+    performing several circular shifts of a shift-variant transform [1]_.
+
+    For a n-level discrete wavelet transforms, one may wish to perform all
+    shifts up to ``max_shifts = 2**n - 1``. In practice, much of the benefit
+    can often be realized with only a small number of shifts per axis.
+
+    For transforms such as the blockwise discrete cosine transform, one may
+    wish to evaluate shifts up to the block size used by the transform.
+
+    References
+    ----------
+    .. [1] R.R. Coifman and D.L. Donoho.  "Translation-Invariant De-Noising".
+           Wavelets and Statistics, Lecture Notes in Statistics, vol.103.
+           Springer, New York, 1995, pp.125-150.
+           DOI:10.1007/978-1-4612-2544-7_9
+
+    Examples
+    --------
+    >>> import skimage.data
+    >>> from skimage import img_as_float
+    >>> from skimage.restoration import denoise_wavelet, cycle_spin
+    >>> img = img_as_float(skimage.data.camera())
+    >>> sigma = 0.1
+    >>> img = img + sigma * np.random.standard_normal(img.shape)
+    >>> denoised = cycle_spin(img, func=denoise_wavelet, max_shifts=3)
+
+    """
+    x = np.asanyarray(x)
+    all_shifts = _generate_shifts(x.ndim, multichannel, max_shifts,
+                                  shift_steps)
+    all_shifts = list(all_shifts)
+    nshifts = len(all_shifts)
+
+    def _run_one_shift(shift):
+        # shift, apply function, inverse shift
+        xs = _roll_axes(x, shift)
+        tmp = func(xs, **func_kw)
+        return _roll_axes(tmp, -np.asarray(shift))
+
+    if not has_futures:
+        if num_workers is None or num_workers > 1:
+            warn("Setting num_workers to 1 because the futures package was not"
+                 "found. On Python 2.7, futures can be installed via:\n"
+                 "pip install futures")
+        num_workers = 1
+
+    avg_y = 0
+    if num_workers == 1:
+        # serial processing
+        for shift in all_shifts:
+            avg_y += _run_one_shift(shift)
+    else:
+        # multithread via concurrent.futures ThreadPoolExecutor
+        if num_workers is None:
+            num_workers = cpu_count()
+        num_workers = np.clip(num_workers, 1, nshifts)
+        with futures.ThreadPoolExecutor(max_workers=num_workers) as execute:
+            for y in execute.map(_run_one_shift, all_shifts):
+                avg_y += y
+    avg_y /= nshifts
```

Comment:
Using a running mean has better numerical stability and (imho) is prettier:

```python
avg_y = np.zeros(x.shape, dtype=float)
for i, y in enumerate(results):
    avg_y += (y - avg_y) / (i + 1)
```

It's a little more expensive but I would expect that in almost all applications, `_run_one_shift` will dominate the runtime anyway.